base/cvd/cuttlefish/host/libs/zip/disjoint_range_set.cc

```diff
@@ -0,0 +1,145 @@
+//
+// Copyright (C) 2025 The Android Open Source Project
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+#include "cuttlefish/host/libs/zip/disjoint_range_set.h"
+
+#include <stdint.h>
+
+#include <algorithm>
+#include <memory>
+#include <set>
+#include <utility>
+#include <vector>
+
+#include <android-base/logging.h>
+
+namespace cuttlefish {
+namespace {
+
+class Range {
+ public:
+  Range(uint64_t start, uint64_t end) : start_(start), end_(end) {}
+
+  bool Contains(const Range& other) const {
+    return start_ <= other.start_ && end_ >= other.end_;
+  }
+
+  bool CanMerge(const Range& other) const {
+    return start_ <= other.end_ && other.start_ <= end_;
+  }
+
+  Range Merge(const Range& other) const {
+    CHECK(CanMerge(other));
+    return Range(std::min(start_, other.start_), std::max(end_, other.end_));
+  }
+
+  bool operator<(const Range& other) const { return start_ < other.start_; }
+  bool operator==(const Range& other) const {
+    return start_ == other.start_ && end_ == other.end_;
+  }
+
+  std::pair<uint64_t, uint64_t> AsPair() const {
+    return std::make_pair(start_, end_);
+  }
+
+ private:
+  uint64_t start_;
+  uint64_t end_;
+};
+
+using Iterator = std::set<Range>::iterator;
+
+}  // namespace
+
+struct DisjointRangeSet::Impl {
+  std::set<Range> ranges_;
```

Comment:
nit: This could be `std::set<const Range>` now to make sure those ranges are not modified while in the set. Modifying the `start_` member is particularly dangerous because it risks breaking the set.

Reply:
I tracked down where this comes from in the standard. In c++14 23.2.1 Table 99 -- Allocator-aware container requirements, it writes:

> ...X denotes an allocator-aware container class with a value_type of T... a and b denote non-const lvalues of type X, t denotes an lvalue or a const rvalue of type X...
>
> - Expression: `a = t`
> - Return type: `X&`
> - Requires: T is CopyInsertable into X and CopyAssignable.
> - post: a == t
> - Complexity: linear